### ISSUE TITLE
the login part as of today doesn't login 

### ISSUE THREAD

**Issue by ohmichael** [winsomehax/plugin.video.bitchute#39]
sorry to bother you again but it seems that when i go to subscriptions or any part of bitchute that it fails to login.

**Comment by winsomehax** [comment]
I just tried it and yeah... "Unable to login with the details provided"

I'll do some digging... my first guess is that bitchute might have changed their interface. As there is no API, the add-on has to use a horrible scraping technique (literally down the page and then take it apart).

**Comment by winsomehax** [comment]
After a bit of digging... it now seems that Bitchute have upgraded/changed some stuff to enforce sending a "User Agent" in the header for any requests. No big deal really... I've made a few additions to the code and updated the "testing" branch. It seems to be working again.

Note - you will need to be using KODI 19 to test it. Let me know if this is a problem

**Comment by ohmichael** [comment]
It might work with python switched

**Comment by ohmichael** [comment]
It might work with python version switched to leia version I was typing from my tv earlier so I redid this- I can’t find where to install it says 2 hours ago but nothing has changed.

**Comment by DavidFW1960** [comment]
> It might work with python version switched to leia version I was typing from my tv earlier so I redid this- I can’t find where to install it says 2 hours ago but nothing has changed.

You need to switch from Master to the Testing branch. There's only 2 files changed. addon.xml and bitchute_access.py so I just overwrote those 2 files with the new ones. Working fine.

**Comment by winsomehax** [comment]
I've now merged this with the master branch as it appears to have fixed the issue. 

Note: if anyone is using a KODI version Pre-19... you can change the line in 
 "addon.xml" that reads:

<import addon="xbmc.python" version="3.0.0"/>

to

<import addon="xbmc.python" version="2.26.0"/>

and try to install it as you normally would. Going forward if I continue to work on this I will end up using more and more Python 3 features and I also have no way to test Kodi <19 versions. That'll mean this trick probably won't work... but for now it may.

For anyone who wasn't aware. Kodi 19 is the version where Kodi dropped support for Python 2 as it's no longer maintained.

**Comment by victoroldschool** [comment]
> > It might work with python version switched to leia version I was typing from my tv earlier so I redid this- I can’t find where to install it says 2 hours ago but nothing has changed.
> 
> You need to switch from Master to the Testing branch. There's only 2 files changed. addon.xml and bitchute_access.py so I just overwrote those 2 files with the new ones. Working fine.

Hey, I'm assuming you're using Kodi v19 (Matrix) and not v18 (Leia)??? I just tried this on a Leia machine and it definitely broke the addon (won't open). I'm assuming it's due to the fact that the new changes are using Python3 which v19 Matrix has, but Leia unfortunately is stuck on Python2.

This is the reason that people are gonna be stuck on Leia for a long time - a ton of addons won't be updated ever (dead development but addon still works), and a lot of other devs are waiting for most users to move that direction before proceeding (because they'll have to maintain TWO versions of the same addon if they wish to support both Kodi versions).

IF you actually did get this working on v18 Leia, please let me know in greater detail how you pulled it off because replacing those two files didn't work (I connected via FTP and simply replaced the two files with the updated ones, as you had mentioned). I would love to get this working again.... I'm disabled and can't really sit at a computer anymore, so this addon was a lifesaver!  :D 




> I've now merged this with the master branch as it appears to have fixed the issue.
> 
> Note: if anyone is using a KODI version Pre-19... you can change the line in
> "addon.xml" that reads:
> 
> import addon="xbmc.python" version="3.0.0"
> 
> to
> 
> import addon="xbmc.python" version="2.26.0"
> 
> and try to install it as you normally would. Going forward if I continue to work on this I will end up using more and more Python 3 features and I also have no way to test Kodi <19 versions. That'll mean this trick probably won't work... but for now it may.
> 
> For anyone who wasn't aware. Kodi 19 is the version where Kodi dropped support for Python 2 as it's no longer maintained.

I'm pretty sure that doesn't work (first thing I tried the other day!). I know XML, html, and some C, but no Python (of course). I'm gonna have to try and learn it and try to port this to Python2 support, as so many users + devs won't be upgrading to v19 Matrix for quite some time due to the hard dependency on Python3 breaking most addons (I wish they would've kept both versions, and simply used Python2 for fallback with addons that haven't been updated yet. Pretty sure I can do that on my Linux machines but I could be mistaken.)

BUT - before I even bother.... from your knowledge and experience as a Python dev - is it even possible to add the required 'user-agent' function to work with Python2??

Super appreciative of all your hard work! I was pleasantly surprised to have seen this topic created and you already replying + making a fix within 24h of it going down.

Take care

**Comment by winsomehax** [comment]
Until very recently this add-on was leia based and could be made to work on
Kodi 19 by changing add-on.xml to ask for python 3.0.0. I had a Kodi 19
branch that did only that for people who wanted to run it on 19 pre release

Literally all I did when 19 was released was make 3.0.0 the default going
forward in add-on.xml

If you get errors on Leia with the master branch after altering add-on.xml
back to python 2... Post them here.

**Comment by victoroldschool** [comment]
> Until very recently this add-on was leia based and could be made to work on Kodi 19 by changing add-on.xml to ask for python 3.0.0. I had a Kodi 19 branch that did only that for people who wanted to run it on 19 pre release Literally all I did when 19 was released was make 3.0.0 the default going forward in add-on.xml If you get errors on Leia with the master branch after altering add-on.xml back to python 2... Post them here.

Well I'll be damned!! IT WORKS!!! I appreciate you helping me out - I will post the exact procedure below for those who might not understand how to do it.

### ***** To get this working on older versions of Kodi, such as v18 'Leia', please do the following: *****

_1._ Download the "**MASTER**" branch of this addon.
_2._ Extract the downloaded addon .zip file on your PC.
_3._ Modify the '**addon.xml**' file with a text editor. Change this specific line:
import addon="xbmc.python" version="3.0.0"

_*Change it to look like this:_
import addon="xbmc.python" version="2.26.0"


_4._ Save + Exit
_5._ We can either connect to the box via FTP and simply copy the addon directory to the kodi addons folder, or we can repackage the zip and install it via the Kodi box itself:

**5A.** _FTP Method:_ Uninstall the old version of Bitchute addon from Kodi (select 'Yes' to erase all addon data/settings). Then connect to the Kodi box via FTP from your PC. Copy the bitchute addon directory to the 'addons' folder. Rename the addon folder to ' plugin.video.bitchute '. Reboot your Kodi box. Login as normal!

**5B.** _Repack ZIP Method:_ Uninstall the old version of Bitchute addon from Kodi (select 'Yes' to erase all addon data/settings). On your PC, we need to create a ZIP of all the files in the addon folder (**NOTE: please lookup elsewhere the correct settings/methods to creating a compatible ZIP for Kodi installations. Make sure the ZIP archive is named ' plugin.video.bitchute '. You can now install it on your Kodi box (with USB stick, NFS, Samba, etc) and login as normal.

Done!

**Comment by winsomehax** [comment]
I've created the KODI-LEIA branch. This contains a copy of the master branch with the addon.xml file modified to require Python 2 - so it can be grabbed as a zip file and installed on pre KODI 19 systems. It also has the user-agent fix.

I will try to fix and issues that come up on this branch too, but future work will focus on 19+ going forward. So they will end up diverging.

**Comment by victoroldschool** [comment]
> I've created the KODI-LEIA branch. This contains a copy of the master branch with the addon.xml file modified to require Python 2 - so it can be grabbed as a zip file and installed on pre KODI 19 systems. It also has the user-agent fix.
> 
> I will try to fix and issues that come up on this branch too, but future work will focus on 19+ going forward. So they will end up diverging.

Oh awesome, thanks so much!! That will definitely make it easier for folks to get it running again. And totally understandable regarding future development, it's more than we could ask for!  :D

The other thing I did do was to change the version number to 2.3.0 (so it wouldn't get over-written by repo updates). I had previously installed the addon from the TVA repo, not direct from Master branch here on GitHub (prob why I couldn't get the mod to work before!).


Anyways, we really do appreciate all the hard work!

Thanks a million!!

**Comment by winsomehax** [comment]
Good point. The version has been set to 2.3.0

**Comment by DavidFW1960** [comment]
Is there a way to install this as a repo so Kodi notifies me the addon has been updated? At the moment I am just downloading the repo as a zip file..

**Comment by DavidFW1960** [comment]
> Good point. The version has been set to 2.3.0

I think you screwed up by committing this to master branch as the addons.xml now specifies python 2 instead of Python 3 - I think you meant to commit this to leia branch only?

**Comment by winsomehax** [comment]
Yeah... my local "master" branch was correct. Dangers of trying to mix/match updates. Fixed.

**Comment by winsomehax** [comment]
> en updated? At the moment I am just downloading the repo as a zip file..

If anyone wants to do the repo stuff I'm happy to accept patches.

**Comment by DavidFW1960** [comment]
> > en updated? At the moment I am just downloading the repo as a zip file..
> 
> If anyone wants to do the repo stuff I'm happy to accept patches.

Ok so I might take a crack at this as it's maybe within my abilities... we'll see.

**Comment by winsomehax** [comment]
Fixed
